File: frontend/src/types/registre.ts
import type { Ecriture } from "./ecriture";

export interface RegistreOption {
  entreprise_id: string;
  categorie: string;
  annee: number;
  mois: number;
  nombre: number;
}

export interface Registre {
  categorie: string;
  entreprise_id: string;
  annee: number;
  mois: number;
  nombre: number;
  total_ht: string;
  total_tva: string;
  total_ttc: string;
  lignes: Ecriture[];
}

export type NatureCompteTva =
  | "collectee"
  | "deductible_charges"
  | "deductible_immobilisations";

export interface TvaCompteDetail {
  compte: string;
  nature: NatureCompteTva | string;
  debit: string;
  credit: string;
  montant_net: string;
}

export interface TvaMensuelle {
  mois: number;
  annee: number;

  tva_collectee: string;
  tva_deductible_charges: string;
  tva_deductible_immobilisations: string;
  tva_deductible: string;
  tva_nette: string;
  tva_a_payer: string;
  credit_tva: string;

  nombre_ecritures: number;
  nombre_lignes_tva: number;
  a_verifier: boolean;
  raisons_verification: string[];
  comptes: TvaCompteDetail[];
}

export interface TvaAnnuelle {
  entreprise_id: string;
  annee: number;
  mensualites: TvaMensuelle[];

  total_tva_collectee: string;
  total_tva_deductible_charges: string;
  total_tva_deductible_immobilisations: string;
  total_tva_deductible: string;
  total_tva_nette: string;

  total_tva_a_payer_technique: string;
  total_credit_tva_technique: string;
  nombre_mois_a_verifier: number;

  source_calcul: string;
  declaration_fiscale_prete: boolean;
  limites: string[];
}

export interface TvaConfiguration {
  id: string;
  cabinet_id: string;
  entreprise_id: string;
  periodicite: "mensuelle" | "trimestrielle" | "autre" | null;
  prorata_applicable: boolean | null;
  prorata_deduction: string | number | null;
  retenue_applicable: boolean | null;
  jour_limite_declaration: number | null;
  delai_saisie_topaze_jours: number | null;
  compte_tva_collectee: string | null;
  compte_tva_recuperable_charges: string | null;
  compte_tva_recuperable_immobilisations: string | null;
  compte_tva_a_payer: string | null;
  compte_credit_tva: string | null;
  notes: string | null;
}

export interface TvaPeriodeV2 {
  id: string;
  entreprise_id: string;
  annee: number;
  mois: number;
  statut: "provisoire" | "validee" | "cloturee" | string;
  tva_collectee: string | number;
  tva_recuperable_charges: string | number;
  tva_recuperable_immobilisations: string | number;
  credit_anterieur: string | number;
  regularisations: string | number;
  retenues_tva: string | number;
  tva_nette: string | number;
  tva_a_payer: string | number;
  credit_a_reporter: string | number;
  a_verifier: boolean;
  anomalies: string[];
  statut_comptable: "calcul_en_cours" | "a_verifier" | "prete_topaze" | "saisie_topaze" | string;
  statut_declaration: "a_preparer" | "a_verifier" | "prete_a_declarer" | "declaree" | "en_retard" | string;
  date_limite_declaration: string | null;
  declared_at: string | null;
  declared_by: string | null;
  declaration_date_reelle: string | null;
  declaration_reference: string | null;
  declaration_note: string | null;
  justificatif_disponible: boolean;
  topaze_entered_at: string | null;
  topaze_entered_by: string | null;
  topaze_batch_reference: string | null;
}

export interface TvaPeriodesAnnee {
  entreprise_id: string;
  annee: number;
  periodes: TvaPeriodeV2[];
  configuration: TvaConfiguration | null;
}

File: backend/app/schemas/registre.py
"""Schémas de sortie du registre comptable et de la TVA comptable."""

import uuid
from decimal import Decimal

from pydantic import BaseModel, ConfigDict

from app.schemas.ecriture import EcritureOut


class RegistreOptionOut(BaseModel):
    """Combinaison de filtres contenant au moins une écriture validée."""

    entreprise_id: uuid.UUID
    categorie: str
    annee: int
    mois: int
    nombre: int


class RegistreOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    categorie: str
    entreprise_id: uuid.UUID
    annee: int
    mois: int

    nombre: int
    total_ht: Decimal
    total_tva: Decimal
    total_ttc: Decimal

    lignes: list[EcritureOut]


class TvaCompteDetailOut(BaseModel):
    compte: str
    nature: str
    debit: str
    credit: str
    montant_net: str


class TvaMensuelle(BaseModel):
    mois: int
    annee: int

    # Compatibilité avec l'ancienne interface.
    tva_collectee: str
    tva_deductible: str
    tva_nette: str
    nombre_ecritures: int

    # Détail comptable V2.
    tva_deductible_charges: str
    tva_deductible_immobilisations: str
    tva_a_payer: str
    credit_tva: str
    nombre_lignes_tva: int
    a_verifier: bool
    raisons_verification: list[str]
    comptes: list[TvaCompteDetailOut]


class TvaAnnuelleOut(BaseModel):
    entreprise_id: uuid.UUID
    annee: int
    mensualites: list[TvaMensuelle]

    total_tva_collectee: str
    total_tva_deductible_charges: str
    total_tva_deductible_immobilisations: str
    total_tva_deductible: str
    total_tva_nette: str

    # Ces deux totaux sont techniques : ils additionnent les positions
    # mensuelles sans appliquer un report fiscal entre périodes.
    total_tva_a_payer_technique: str
    total_credit_tva_technique: str

    nombre_mois_a_verifier: int
    source_calcul: str
    declaration_fiscale_prete: bool
    limites: list[str]
